### PR TITLE
New FPGA code sample "ac_fixed"

### PR DESCRIPTION
**Pull request by wangdi4**
# Adding a New Sample "ac_fixed"

## Description

This FPGA tutorial shows how to use the `ac_fixed` data type with some simple examples.

## Checklist
Administrative
- [x] Review sample design with the appropriate [Domain Expert](https://github.com/oneapi-src/oneAPI-samples/wiki/Reviewers-and-Domain-Experts): Tanner (@tyoungsc)

Code Development
- [x] Implement coding guidelines and ensure code quality. [see wiki for details](https://github.com/oneapi-src/oneAPI-samples/wiki/General-Code-Guidelines)
- [x] Adhere to readme template 
- [x] Enforce format via clang-format config file
- [x] Adhere to sample.json specification. https://github.com/oneapi-src/oneAPI-samples/wiki/sample-json-specification
- [x] Ensure/create CI test configurations for sample (ciTests field) https://github.com/oneapi-src/oneAPI-samples/wiki/sample-json-ci-test-object
- [x] Run jsonlint on sample.json to verify json syntax. www.jsonlint.com

Security and Legal
- [ ] OSPDT Approval (see Project Manager for assistance)
- [ ] Compile using the following compiler flags and fix any warnings, the falgs are: "/Wall -Wformat-security -Werror=format-security"
- [ ] Virus scan

Review
- [ ] Review DPC++ code with Paul Peterseon. (GitHub User: pmpeter1)
- [ ] Review readme with Tom Lenth(@tomlenth) and/or Project Manager
- [ ] Tested using Dev Cloud when applicable

### CODE REVIEW

**Comment by tyoungsc** [comment]
@wangdi4 I noticed you did a merge of master into your feature branch. In the future, try to use `git rebase` rather than `git merge`.

File: DirectProgramming/DPC++FPGA/Tutorials/Features/ac_fixed/src/ac_fixed.cpp
Comment (keryell) [inline DirectProgramming/DPC++FPGA/Tutorials/Features/ac_fixed/src/ac_fixed.cpp:126]: What are all these `AC_RND` and `AC_SAT`? Where do they come from? Some macros defined somewhere?

File: DirectProgramming/DPC++FPGA/Tutorials/Features/ac_fixed/src/ac_fixed.cpp
Comment (pmpeter1) [inline DirectProgramming/DPC++FPGA/Tutorials/Features/ac_fixed/src/ac_fixed.cpp:72]: For better readability would using sin_x (instead of sin_ret) and cos_x (instead of cos_ret) work better?

File: DirectProgramming/DPC++FPGA/Tutorials/Features/ac_fixed/src/ac_fixed.cpp
Comment (keryell) [reply DirectProgramming/DPC++FPGA/Tutorials/Features/ac_fixed/src/ac_fixed.cpp:126]: The problem is to understand where they come from.
Is `ac_fixed` defined in `::`? Same for `AC_RND` and `AC_SAT` or are they macros (argh!)? It does not fit well in a good looking C++ program.

**Review by tomlenth — APPROVED** [review]

File: DirectProgramming/DPC++FPGA/Tutorials/Features/ac_fixed/src/ac_fixed.cpp
Comment (tyoungsc) [reply DirectProgramming/DPC++FPGA/Tutorials/Features/ac_fixed/src/ac_fixed.cpp:125]: `ac_fixed` is defined in `::`.
Looks like`AC_RND` and `AC_SAT` are enums in `::` (from ac_int.hpp around line 1234):

```cpp
enum ac_q_mode {
  AC_TRN,
  AC_RND,
  AC_TRN_ZERO,
  AC_RND_ZERO,
  AC_RND_INF,
  AC_RND_MIN_INF,
  AC_RND_CONV,
  AC_RND_CONV_ODD
};
```

A decent change may be moving from an `enum` to an `enum class`, but these files are from Mentor Graphics, so we many need to take up any changes with them. For now, it may be worthwhile adding to the comment where these values are coming from.

Comment (keryell) [reply DirectProgramming/DPC++FPGA/Tutorials/Features/ac_fixed/src/ac_fixed.cpp:125]: Yes, and having `ac::trn` instead of `AC_TRN`, but this is another battle to have. :-)

Comment (tyoungsc) [reply DirectProgramming/DPC++FPGA/Tutorials/Features/ac_fixed/src/ac_fixed.cpp:125]: agreed :|

**Review by pmpeter1 — DISMISSED** [review]

File: DirectProgramming/DPC++FPGA/Tutorials/Features/ac_fixed/README.md
Comment (yuguen-intel) [inline DirectProgramming/DPC++FPGA/Tutorials/Features/ac_fixed/README.md:89]: Are they also documented in the documentation linked above? Where is a user going to find this file?

File: DirectProgramming/DPC++FPGA/Tutorials/Features/ac_fixed/README.md
Comment (yuguen-intel) [inline DirectProgramming/DPC++FPGA/Tutorials/Features/ac_fixed/README.md:93]: Same remark as above

File: DirectProgramming/DPC++FPGA/Tutorials/Features/ac_fixed/README.md
Comment (yuguen-intel) [inline DirectProgramming/DPC++FPGA/Tutorials/Features/ac_fixed/README.md:98]: Please precise that this is the case here because of the fixed-point datatype chosen for this example. One could have chosen a different ac_fixed datatype that would have resulted in a more accurate than floating point result

File: DirectProgramming/DPC++FPGA/Tutorials/Features/ac_fixed/README.md
Comment (yuguen-intel) [inline DirectProgramming/DPC++FPGA/Tutorials/Features/ac_fixed/README.md:266]: What are these "Constructed from" refer to? is this the result? If so, as the output is identical it is hard to understand below why there is a "MAX DIFF".
You can increase the number of digits to be displayed in order to show the difference between the two numbers.

File: DirectProgramming/DPC++FPGA/Tutorials/Features/ac_fixed/README.md
Comment (yuguen-intel) [inline DirectProgramming/DPC++FPGA/Tutorials/Features/ac_fixed/README.md:269]: You could add a message saying: "computing on <x> values between <min> and <max>" followed by a message (not in full caps as none of the other messages are in full caps) saying "maximum difference observed between the golden value and the computed value" 
Also what is the methodology for knowing what is the golden value?

File: DirectProgramming/DPC++FPGA/Tutorials/Features/ac_fixed/README.md
Comment (yuguen-intel) [inline DirectProgramming/DPC++FPGA/Tutorials/Features/ac_fixed/README.md:284]: For all these results, I don't know from this what they refer to, please add more print messages explaining what is beeing computed

File: DirectProgramming/DPC++FPGA/Tutorials/Features/ac_fixed/README.md
Comment (yuguen-intel) [inline DirectProgramming/DPC++FPGA/Tutorials/Features/ac_fixed/README.md:291]: Please add a comment saying that this is the case because the conversion will occur on the host rather than on the compute kernel.
And I would rephrase the second sentence as it suggests that the ac_fixed is less acccurate and more resource efficient no matter what which is not the case. I can create a fixed point format so large it covers the entire range of the floating-point format. In that case the resource utilisation is going to be way worse for exactly the same accuracy.
What is hard when using fixed-point numbers is knowing if your application could benefit from a fixed-point format, which is usually not the case.

File: DirectProgramming/DPC++FPGA/Tutorials/Features/ac_fixed/ac_fixed.vcxproj
Comment (yuguen-intel) [inline DirectProgramming/DPC++FPGA/Tutorials/Features/ac_fixed/ac_fixed.vcxproj:1]: Have you tested these MSVS files?

File: DirectProgramming/DPC++FPGA/Tutorials/Features/ac_fixed/src/CMakeLists.txt
Comment (yuguen-intel) [inline DirectProgramming/DPC++FPGA/Tutorials/Features/ac_fixed/src/CMakeLists.txt:38]: This looks different than the other CMakeLists.txt files - do we need to specify here that we don't need the linker flag when building the report? Can we just instead use the hardware link flags? Just to keep all the cmake files as close as possible to one another

File: DirectProgramming/DPC++FPGA/Tutorials/Features/ac_fixed/src/ac_fixed.cpp
Comment (yuguen-intel) [inline DirectProgramming/DPC++FPGA/Tutorials/Features/ac_fixed/src/ac_fixed.cpp:6]: Why do we need these clang-format on/off comments here? 
I don't see us using these in the ac_int tutorial for example

File: DirectProgramming/DPC++FPGA/Tutorials/Features/ac_fixed/src/ac_fixed.cpp
Comment (yuguen-intel) [inline DirectProgramming/DPC++FPGA/Tutorials/Features/ac_fixed/src/ac_fixed.cpp:29]: You made two type aliases above (fixed_10_3_t and fixed_9_2_t), why don't you have a third one with this datatype? It could also be a good to explain in this unique location what AC_RND and AC_SAT parameters are doing

File: DirectProgramming/DPC++FPGA/Tutorials/Features/ac_fixed/src/ac_fixed.cpp
Comment (yuguen-intel) [inline DirectProgramming/DPC++FPGA/Tutorials/Features/ac_fixed/src/ac_fixed.cpp:27]: Here the variable that holds the result is called b, but is called ret in TestConstructFromACFixed

File: DirectProgramming/DPC++FPGA/Tutorials/Features/ac_fixed/src/ac_fixed.cpp
Comment (yuguen-intel) [inline DirectProgramming/DPC++FPGA/Tutorials/Features/ac_fixed/src/ac_fixed.cpp:63]: Here the input is called x where it is called a in the two above functions

File: DirectProgramming/DPC++FPGA/Tutorials/Features/ac_fixed/src/ac_fixed.cpp
Comment (yuguen-intel) [inline DirectProgramming/DPC++FPGA/Tutorials/Features/ac_fixed/src/ac_fixed.cpp:90]: Why do you need clang-format on/off here?

File: DirectProgramming/DPC++FPGA/Tutorials/Features/ac_fixed/src/ac_fixed.cpp
Comment (yuguen-intel) [inline DirectProgramming/DPC++FPGA/Tutorials/Features/ac_fixed/src/ac_fixed.cpp:91]: Input called x vs a in above functions, please use the same convention for all these functions

File: DirectProgramming/DPC++FPGA/Tutorials/Features/ac_fixed/README.md
Comment (wangdi4) [reply DirectProgramming/DPC++FPGA/Tutorials/Features/ac_fixed/README.md:46]: It is now. Are you concerning about the version update of the document in the future?

File: DirectProgramming/DPC++FPGA/Tutorials/Features/ac_fixed/README.md
Comment (wangdi4) [reply DirectProgramming/DPC++FPGA/Tutorials/Features/ac_fixed/README.md:89]: The document is in oneAPI optimization guide:  https://www.intel.com/content/www/us/en/develop/documentation/oneapi-fpga-optimization-guide/top/optimize-your-design/resource-use/data-types-and-operations/var-prec-fp-sup/declare-and-use-the-ac-data-types/declare-the-ac-fixed-data-type.html,

which also refers to the `ac_fixed_math.hpp` file. Should we just mention this page in the tutorial or directly refer to the header file?

File: DirectProgramming/DPC++FPGA/Tutorials/Features/ac_fixed/README.md
Comment (wangdi4) [reply DirectProgramming/DPC++FPGA/Tutorials/Features/ac_fixed/README.md:269]: The `MAX DIFF` here is actually quantum instead of difference between values. Maybe we should also display the difference between the result and the golden value.

The golden value can be hardcoded

File: DirectProgramming/DPC++FPGA/Tutorials/Features/ac_fixed/README.md
Comment (yuguen-intel) [inline DirectProgramming/DPC++FPGA/Tutorials/Features/ac_fixed/README.md:89]: files -> file

File: DirectProgramming/DPC++FPGA/Tutorials/Features/ac_fixed/README.md
Comment (yuguen-intel) [inline DirectProgramming/DPC++FPGA/Tutorials/Features/ac_fixed/README.md:308]: "of a smaller size" is not quite correct as you can use a larger fixed-point format than float and have arithmetic operators that use fewer resources than a floating point operator.
I would rephrase as follows: 
 "Additionally, by using the `ac_fixed` types and math library functions, you can use an even smaller fixed point format to trade-off even more accuracy for a more resource efficient design if your application requirements allows for it."

File: DirectProgramming/DPC++FPGA/Tutorials/Features/ac_fixed/src/ac_fixed.cpp
Comment (yuguen-intel) [inline DirectProgramming/DPC++FPGA/Tutorials/Features/ac_fixed/src/ac_fixed.cpp:19]: Can you say here where a user can find the list of all the modes available?

**Review by yuguen-intel — CHANGES_REQUESTED** [review]

**Review by yuguen-intel — APPROVED** [review]
Looks good to me, approving to get the CI flow running

**Comment by yuguen-intel** [comment]
The CI flow is currently disabled - I did the validation manualy. 
It is only missing @tomlenth and @pmpeter1 approval before merging

**Pull request merged**
